mod_papers/helper.php

```diff
@@ -13,150 +14,257 @@
-                        if (($authors === reset($work['work-contributors']['contributor'])) && ($authors === end($work['work-contributors']['contributor']))) {
-                            $output .= $authors['credit-name']['value'] . '<br>';
-                        } elseif ($authors === reset($work['work-contributors']['contributor'])) {
-                            $output .= $authors['credit-name']['value'];
-                        } elseif ($authors === end($work['work-contributors']['contributor'])) {
-                            $output .= ' and ' . $authors['credit-name']['value'] . '<br>';
-                        } else {
-                            $output .= ', ' . $authors['credit-name']['value'];
-                        }
-                    }
-                } else {
-                    //Get authorlist from bibtex
-                    if (preg_match('/author\\s?=\\s?{(.+)}/', $bibtex, $match)) {
-                        $authorstr = $match[1];
-                        $authors   = explode(" and ", $authorstr);
-                        foreach ($authors as $author) {
-                            if (($author === reset($authors)) && ($author === end($authors))) {
-                                $output .= $author . '<br>';
-                            } elseif ($author === reset($authors)) {
-                                $output .= $author;
-                            } elseif ($author === end($authors)) {
-                                $output .= ' and ' . $author . '<br>';
-                            } else {
-                                $output .= ', ' . $author;
-                            }
-                        }
-                    }
-                }
-
-                $output .= '<a href="http://dx.doi.org/' . $work['udoi'] . '">' . $work['journal-title']['value'] . ' <b>' . $volume . '</b> ' . $pages . ' (' . $work['publication-date']['year']['value'] . ')</a><br style="line-height:2.5em;">';
-            }
-        }
-
-        return $output;
-    }
+	/**
+	 * Retrieves the required informaton from ORCID
+	 *
+	 * @param   array  $orcids An object containing the ORCIDs to be searched
+	 *
+	 * @access public
+	 */
+
+	public static function getPapers($orcids){
+		$lastRunLog =  dirname(__FILE__) . '/lastrun.log';
+		$data_file = dirname(__FILE__) . '/pubs.html';
+		if (file_exists($lastRunLog) and file_exists($data_file)) {
+			$lastRun = file_get_contents($lastRunLog);
+			if (time() - (int)$lastRun >= 86400) {
+				//its been more than a day so run our external file
+				$papers = self::genPapers($orcids);
+				//update lastrun.log with current time
+				file_put_contents($lastRunLog, time());
+			} else {
+				$papers = file_get_contents($data_file);	
+			}
+		} else {
+			$papers = self::genPapers($orcids);
+		}
+
+		return $papers;
+	}
+
+	public static function genPapers($orcids)
+	{
+		// array of curl handles
+		$multiCurl = array();
```

Comment:
I wasn't aware of multi-curl! I'll take this portion at face value. I assume though that our `foreach ($orcids as $id)` loop no longer is called sequentially, but instead we send off 3 requests simultaneously? If so: that's cool.

Reply:
Yes, it sends three requests in parallel as far as I know. There is still a problem - some requests end ups with null.  The situation has been somehow improved after adding usleep() calls. But still I have to learn more about setting multi-curl up - it has dozens of parameters.